### PR TITLE
8309460: JScrollBar leaves behind clutter for non-integer UI scales

### PR DESCRIPTION
**Pull request by prsadhuk**
Lines  are left behind when moving the scrollbar in the positive direction. but are cleaned up on mouse release. 
Additonally, with right arrow clicks too, the lines are left behind.
Seems like for mouseDragged and scrollByUnit, the dirty region of the scrollbar is not repainted leading to artifacts
which is being done in this fix..

---------
### Progress
- [ ] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8309460](https://bugs.openjdk.org/browse/JDK-8309460): JScrollBar leaves behind clutter for non-integer UI scales (**Bug** - P3)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/17484/head:pull/17484` \
`$ git checkout pull/17484`

Update a local copy of the PR: \
`$ git checkout pull/17484` \
`$ git pull https://git.openjdk.org/jdk.git pull/17484/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 17484`

View PR using the GUI difftool: \
`$ git pr show -t 17484`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/17484.diff">https://git.openjdk.org/jdk/pull/17484.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/17484#issuecomment-1898298219)

### CODE REVIEW

**Comment by mrserb** [comment]
Why the bug is reproduced only if non-integer UI is used?

**Comment by prsadhuk** [comment]
I/m not sure but in 2 cases where it cannot be reproduced ie during mouseRelease and block increment, they do the dirty region painting which I guess should be done here too irrespectively..
https://github.com/openjdk/jdk/blob/9efdd242fb40a8270e489cc071ff1c891878e24f/src/java.desktop/share/classes/javax/swing/plaf/basic/BasicScrollBarUI.java#L1255-L1256
and
https://github.com/openjdk/jdk/blob/9efdd242fb40a8270e489cc071ff1c891878e24f/src/java.desktop/share/classes/javax/swing/plaf/basic/BasicScrollBarUI.java#L1122-L1123

File: test/jdk/javax/swing/JScrollBar/JScrollBarArtifactTest.java [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+import java.awt.BorderLayout;
+import javax.swing.JFrame;
+import javax.swing.JPanel;
+import javax.swing.JScrollBar;
+import javax.swing.SwingUtilities;
+
+/*
+ * @test
+ * @bug 8309460
+ * @summary Verifies clicking JComboBox during frame closure causes Exception
```
Comment (honkar-jdk) [inline test/jdk/javax/swing/JScrollBar/JScrollBarArtifactTest.java:33]: summary needs to be updated

**Comment by mrserb** [comment]
> Also, it might be the case the scrollbar methods like getUnitIncrement, scrollBar.getValue scrollbar.setValue setThumbBounds sets or returns or uses integer value despite scale being non-integer, so I guess repainting dirty region should be done to refresh the artifacts

But in the fix and in the places you pointed out we also use the int values, so why is the bug reproduced only when the float scale is used?

**Review by honkar-jdk — COMMENTED** [review]
The fix works fine. The artifacts or lines are no longer visible.
But noticed a couple of things when testing. (Win 11)

1. The lines happen to occur for **non-standard fractional scales** (2.3, 2.7) but work fine for **standard fractional scale** supported by windows (1.25,1.75, 2.25...)

2. The lines are left behind only in one direction - right for horizontal scroll and down for vertical scroll.

3. With the present fix repaint is called each time mouse is dragged or clicked, so that results in a lot of calls to repaint during scrolling action.

**Comment by aivanov-jdk** [comment]
> > Also, it might be the case the scrollbar methods like getUnitIncrement, scrollBar.getValue scrollbar.setValue setThumbBounds sets or returns or uses integer value despite scale being non-integer, so I guess repainting dirty region should be done to refresh the artifacts
> 
> But in the fix and in the places you pointed out we also use the int values, so why is the bug reproduced only when the float scale is used?

Because float scale results in fractional pixels?

If the scale is 1.25, then 1 pixel is 1.25 physical pixels; 2 pixels are 2.5 physical pixels and so on.

File: src/java.desktop/share/classes/javax/swing/plaf/basic/BasicScrollBarUI.java [hunk 2]
```diff
@@ -1188,6 +1188,8 @@ else if (delta < 0 && newValue > oldValue) {
      */
     protected void scrollByUnit(int direction)  {
         scrollByUnits(scrollbar, direction, 1, false);
+        Rectangle r = getTrackBounds();
+        scrollbar.repaint(r.x, r.y, r.width, r.height);
```
Comment (aivanov-jdk) [inline src/java.desktop/share/classes/javax/swing/plaf/basic/BasicScrollBarUI.java:1192]:
```suggestion
        scrollbar.repaint(getTrackBounds());
```

**Comment by honkar-jdk** [comment]
> Because float scale results in fractional pixels?
> 
> If the scale is 1.25, then 1 pixel is 1.25 physical pixels; 2 pixels are 2.5 physical pixels and so on.

Initially this issue looked similar to fractional scaling issues such as border rendering issues observed on windows previously but it seems to happen only on non-standard fractional UI scales such as 2.3, 2.7 and not on standard fractional scales supported on windows (1.25,1.75,2.25 ..) https://github.com/openjdk/jdk/pull/17484#pullrequestreview-1834275178.

**Comment by prsadhuk** [comment]
Looks to be related to 
8032219: [macosx] Scrollbars looks bad under retina in Motif and Metal L&F fix 
which changes the way scrollbar is drawn for Metal and Motif L&F by using fillRect instead of drawLine

and the current issue also happens for Metal and Motif L&F
but not for say, Windows and Nimbus L&F

**Comment by honkar-jdk** [comment]
@prsadhuk @aivanov-jdk With the original solution the scrollbar repaint is done only when mouse is released or scrolled by a block, with the proposed fix the repaint is called continuously as long as the mouse is dragged or scrolled by a unit. Won't this create a lot of calls to repaint and thus cause a overhead?

**Comment by aivanov-jdk** [comment]
> @prsadhuk @aivanov-jdk I have a clarification regarding the proposed fix.
> 
> With the original solution the scrollbar repaint is done only when mouse is released or scrolled by a block, with the proposed fix the repaint is called continuously as long as the mouse is dragged or scrolled by a unit. Won't this result in a lot of calls to repaint and thus cause an overhead?

Additional repaints can create an additional overhead. On the other hand, repaint requests are coalesced. If the scrollbar wasn't repainted, you wouldn't see the thumb move, so there are repaints.

What I don't understand is why those repaints aren't enough and adding one more, apparently at the same location, solves the problem.

When you drag the thumb, to paint it in the new location, you need to paint the background over the old location and then the thumb in the new location. The suggested fix adds one more repaint of the rectangle where the scroll thumb is currently located. Or maybe not, and `getTrackBounds` returns the new rectangle by the time it's called in `scrollByUnit`.

The call to `scrollByUnits` one line above updates the location of the thumb, which changes the model and eventually should call repaint for the old position of the thumb and for the new position of the thumb. The dragging code could handle repaints in an optimised way though.

**Comment by aivanov-jdk** [comment]
> > Because float scale results in fractional pixels?
> > If the scale is 1.25, then 1 pixel is 1.25 physical pixels; 2 pixels are 2.5 physical pixels and so on.
> 
> Initially this issue looked similar to fractional scaling issues such as border rendering issues observed on windows previously but it seems to happen only on non-standard fractional UI scales such as 2.3, 2.7 and not on standard fractional scales supported on windows (1.25,1.75,2.25 ..) [#17484 (review)](https://github.com/openjdk/jdk/pull/17484#pullrequestreview-1834275178).

The underlying problem is still the same: there are fractional pixels.

I haven't run the test case myself. The attached screenshot has a pattern… not every pixel remains visible. Likely, these pixels are where the fractional pixels add up to produce a new full pixel.

**Comment by honkar-jdk** [comment]
> The suggested fix adds one more repaint of the rectangle where the scroll thumb is currently located. Or maybe not, and `getTrackBounds` returns the new rectangle by the time it's called in `scrollByUnit`.

This seems like a possible cause.

> I wonder if the problem can be reproduced by programmatically moving the thumb instead of dragging it with the mouse. Playing with the window size, uiScale, and the max value of the scrollbar could be required…

Good idea. I agree especially the case of max value of scroll bar.

**Comment by prsadhuk** [comment]
> The suggested fix adds one more repaint of the rectangle where the scroll thumb is currently located. Or maybe not, and `getTrackBounds` returns the new rectangle by the time it's called in `scrollByUnit`.

`getTrackingBounds` returns the whole scrolling area bounds, in my case it is  java.awt.Rectangle[x=15,y=0,width=755,height=17]
each time `scrollByUnit `is called so it is repainting the whole scrolling area

**Comment by mrserb** [comment]
>Is this bug similar to [JDK-8032219](https://bugs.openjdk.org/browse/JDK-8032219)?

That bug was caused by mixing the vector-based API(lines) and rasters(fillRect). At that time we rendered the line outside of the component's bounds, and that part of the line was not cleared by the fillRect. So the fix changed how we render the lines(borders), to always render them inside the component.

But in this case, it is unclear what is the actual problem. How it related to the "non-standard" fractional metrics(such as 2.3, 2.7) and why it works fine for 1.5 and 1.75. If we draw a border outside of the component that border should be fixed. If the clear/fillRect does not clear the component then we should fix the calculation of the bounds.

Note the proposed patch just repaints the component using its bounds, so I assume the bounds are correct, the line is properly rendered inside of the component, and something different is broken.

**Comment by aivanov-jdk** [comment]
> > > Because float scale results in fractional pixels?  
> > > If the scale is 1.25, then 1 pixel is 1.25 physical pixels; 2 pixels are 2.5 physical pixels and so on.
> > 
> > Initially this issue looked similar to fractional scaling issues such as border rendering issues observed on windows previously but it seems to happen only on non-standard fractional UI scales such as 2.3, 2.7 and not on standard fractional scales supported on windows (1.25,1.75,2.25 ..) [#17484 (review)](https://github.com/openjdk/jdk/pull/17484#pullrequestreview-1834275178).
> 
> The underlying problem is still the same: there are fractional pixels.

I ran the test now and I'm sure the problem is very similar. When drag the thumb, the width of the left border is usually 2 pixels, sometimes it changes to 3 pixels. As I move farther, that third pixel remains on the background of the scrollbar.

The left over pixels create a regular pattern: 17 pixels followed by 8 pixels, then the cycle repeats. These are physical display pixels.

If you convert user space pixels to physical pixels, you'll see there's a pattern where fraction adds up to whole pixel: 4 user pixels, 3 user pixels.

As you drag the thumb, you can see that the highlight/shadow lines on the thumb itself dance.

This means that as you drag the thumb, the app sees 1-pixel increments which results in 2-pixel and 3-pixel increments on the screen.


> How it related to the "non-standard" fractional metrics(such as 2.3, 2.7) and why it works fine for 1.5 and 1.75.

I believe I explained how it's related to fractional scales. Why does it work for 1.25 and 1.75? I guess we never end up in a situation where a physical pixel doesn't get painted. With 1.5, the probability is even less.

You can still see the width of the left (and right) border changes; the highlight and shadow lines on the thumb continue to dance.

There are this kind of problems wherever pixels are painted by 1-pixel lines. Look at the title bar of a JInternalFrame in SwingSet2 as drag it: the pattern changes even at 1.5 scale; dragging at 1.25 and 1.75 gives another pattern.

> If we draw a border outside of the component that border should be fixed. If the clear/fillRect does not clear the component then we should fix the calculation of the bounds.

No, we don't draw outside of component bounds. We just don't repaint enough pixels as thumb is moved.

As you rightfully noticed something different is needed. Repainting the bounds of the thumb still solves the problem but it doesn't seem right.

As the thumb is dragged, it's repainted in the new location; the old location is erased eventually. When erasure is handled, I'd adjust the left coordinate to be (left - 1) to repaint a larger rectangle or something like this.

**Comment by aivanov-jdk** [comment]
> The test can be converted to automatic by comparing the captured scroll area before MousePress and between MousePress and Release events.

I agree, the test can be automatic. Grab the thumb with robot and move to the right edge. Without releasing the left mouse button¹, create a multiresolution screenshot to get all the pixels without scaling the image down. The background of the scrollbar on the captured screenshot should not contain darker pixels.

¹ After you release the mouse button, the position of the thumb is adjusted, and the entire scrollbar is repainted. At which point, the leftover darker lines disappear from the scrollbar background.

**Pull request closed without merging**